### PR TITLE
Feature/issue 551 threaded ad

### PR DESCRIPTION
**Pull request by wds15**
#### Submission Checklist

- [ ] Run unit tests: `./runTests.py test/unit`
- [X] Run cpplint: `make cpplint`
- [X] Declare copyright holder and open-source license: see below

#### Summary:
Makes the AD tape **optionally** thread local such that AD can be performed within threads. This will enable threaded use of stan-math and pave the way for a threaded version of `map_rect`. Threading will only be turned on if the `STAN_THREADS` compile time variable is defined.

This change makes  **optionally** the AD tape `thread_local` which is a standard C++11 facility. However, in order to make this change work with g++ and clang++ I had to use a design which resembles what is done in the clang sources (see [context.h](https://github.com/llvm-mirror/clang-tools-extra/blob/master/clangd/Context.h#L74) and [context.cpp](https://github.com/llvm-mirror/clang-tools-extra/blob/master/clangd/Context.cpp#L24) from a clang project). The technique used there relies on a static thread_local variable declared inside a static function. All this function does is to return a reference to an instance of the stack. This function is called `context()` [(see line 15 of `autodiffstackstorage.hpp`)](https://github.com/stan-dev/math/blob/feature/issue-551-threaded-AD/stan/math/rev/core/autodiffstackstorage.hpp#L15).

As a consequence our codebase has to access the stack differently. So far we prefixed everything with `ChainableStack::` which now becomes `ChainableStack::context().`. For example, this

```cpp
// old
ChainableStack::var_stack_.push_back(this);
```

now becomes

```cpp
// new
ChainableStack::context().var_stack_.push_back(this);
```

What this means for `stan-math` users (how to enable & when it works) is documented [on our wiki](https://github.com/stan-dev/math/wiki/Threading-Support).

#### Related discussions
ChainableStack wrapper - http://discourse.mc-stan.org/t/wrapping-calls-to-chainablestack/3779
Dropping support for older Apple Clang releases - http://discourse.mc-stan.org/t/potentially-dropping-support-for-older-versions-of-apples-version-of-clang/3780

#### Intended Effect:
Make AD thread safe. Enable threaded use of AD to allow client code to run AD in threads safely.

#### How to Verify:
I have added a threaded evaluation of the `gradient` facility to the test file `test/unit/math/rev/mat/functor/gradient_test.cpp` which verifies that this runs in parallel. To enable the threading please test with `STAN_THREADS` defined at compile-time.

#### Side Effects:
None.

At least I have not observed any performance impact or so. That is, there is no additional overhead at all as it seems. The compiler appears to just compile away the additional function call.

#### Copyright and Licensing

Please list the copyright holder for the work you are submitting (this will be you or your assignee, such as a university or company): Sebastian Weber

By submitting this pull request, the copyright holder is agreeing to license the submitted work under the following licenses:
- Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
- Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

### CODE REVIEW

**Comment by wds15** [comment]
The `clang++` on the Jenkins complains with

```
./stan/math/rev/core/autodiffstackstorage.hpp:16:12: error: thread-local storage is not supported for the current target
```

which I don't quite understand. On my machine this just compiles fine using Apple's clang or the opensource 5.0 or 6.0 version - all are OK (g++ 6.4 also works).

Does the `clang++` on the Jenkins box need to be reconfigured? The `thread_local` thing is part of the C++11 standard - so this should work as long as we configure the compiler right. From the clang [website](https://clang.llvm.org/cxx_status.html) it quotes as only requirement for `thread_local` to work that we have:

```
(5): thread_local support requires a C++ runtime library providing __cxa_thread_atexit, such as libc++abi 3.6 or later, or libsupc++ 4.8 or later.
```

**Comment by wds15** [comment]
for the record. We are likely hit on Linux/clang++ by this issue:

https://stackoverflow.com/questions/29322666/undefined-reference-to-cxa-thread-atexitcxxabi-when-compiling-with-libc#30437761

**Comment by wds15** [comment]
ok, travis build on linux using clang seems to work now. The changes to the `make/tests` file will probably need to be revise; I did not know where to put this otherwise.

**Comment by seantalts** [comment]
This seems like a good thing for me to review - if no one minds I'll start tonight.

**Comment by wds15** [comment]
Thanks much @seantalts for taking this on. Your shiny new performance testsuite will be of much use for this one. If possible we should make sure that under clang and g++ the performance does not change. BTW, is there a wiki page describing how one can use this testsuite? I am keen to try it out myself.

Right now I am not too happy with how things are under Linux + clang. What I have done is to "make it work", but I hope there is a more elegant solution. As I do not have access to this combination (clang on Linux) it is rather hard for me to find a good solution.

**Comment by seantalts** [comment]
Hey, I'm just getting started looking through this, but for now I put up some instructions for the performance test tools I've been working on, at the README for its new repo: github.com/stan-dev/performance-tests-cmdstan

**Comment by wds15** [comment]
Thanks for the readme. This looks very much like a tool I always wanted as it takes away this tedious juggling of branches & benchmarks. Cool.

**Comment by seantalts** [comment]
Hey @danluu, if you have a second and want to give your $0.02 on the way we're adding C++11 multithreading with this PR that'd be awesome. Curious if there's any better way to do cross-compiler support especially.

File: lib/gtest_1.7.0/src/tls-linux-clang.cpp [hunk 1]
```diff
@@ -0,0 +1,6 @@
+// needed on linux when using clang for thread_local storage to work
```
Comment (seantalts) [inline lib/gtest_1.7.0/src/tls-linux-clang.cpp]: Can you include links in here to the stack overflow stuff just for future reference?

File: test/unit/math/rev/mat/functor/gradient_test.cpp [hunk 2]
```diff
@@ -26,6 +31,51 @@ TEST(AgradAutoDiff, gradient) {
...
+    stan::math::gradient(fun1(), x, fx, grad_fx);
+    VectorXd res(1 + grad_fx.size());
+    res(0) = fx;
+    res.tail(grad_fx.size()) = grad_fx;
+    return res;
+  };
+
+  std::vector<std::future<VectorXd>> ad_futures;
+
+  for (std::size_t i = 0; i < 100; i++) {
```
Comment (seantalts) [inline test/unit/math/rev/mat/functor/gradient_test.cpp]: I think it'd be good to give the test some opportunities to mess up - right now every thread is running the same function with the exact same inputs and outputs. Can you add some work in there and some dependence on `i` maybe?

**Comment by seantalts** [comment]
Re: Jenkins Clang - it's an oldish version of Apple's clang, but I believe that's currently a compiler we're interested in supporting, as far as I know. This has some details but no fixes: https://stackoverflow.com/questions/23791060/c-thread-local-storage-clang-503-0-40-mac-osx

Maybe we should stop supporting it and recommending people install clang from Xcode?

**Comment by bob-carpenter** [comment]
Isn't latest Xcode the current recommendation for Stan on Macs?

I think it'd be OK to stop supporting old apple versions.

I know there was some discussion of trying to put together our own installs with later compilers (not earlier ones).

- Bob

File: make/os_linux [hunk 3]
```diff
@@ -8,6 +8,7 @@
 
 ifeq (g++,$(CC_TYPE))
   LDFLAGS += -lpthread
+  CXXFLAGS += -pthread
```
[the remark this answers is not in the document]
Comment (wds15) [reply make/os_linux]: Correct.

File: lib/gtest_1.7.0/src/tls-linux-clang.cpp (hunk 1 above)
Comment (wds15) [reply lib/gtest_1.7.0/src/tls-linux-clang.cpp]: Sure

**Comment by wds15** [comment]
argh... clang on the Jenkins mac doesn't even understand the `-femulated-tls` flag.

It looks like we cannot support the older compilers on Apple if `thread_local` is in the code base. As I recall Apple has announced the `thread_local` support on WWDC 2016 such that I would anticipate that you need at least Sierra for a supported `thread_local` feature.

My strong preference would be to go ahead with `thread_local` being a required supported feature of the compiler (after all it is plain C++11). Should we discuss at the meeting tomorrow? As long as we can demonstrate that the `thread_local` thing does not cost us any noticeable performance we should just use it and do not add another if-def flag to the code which cuts-out threading. In short: Should `stan-math` require `thread_local` ready compilers?

**Comment by wds15** [comment]
and using libstdc++ on linux also does not help. I think I better stop bombarding the test machines... can someone look into the clang+linux issue if possible? There should be a better solution than this hack which we would need to propagate to the interfaces. @seantalts or @bgoodri maybe? That would be great! Otherwise I would have to install a virtual machine for this.

**Comment by seantalts** [comment]
Moving discussion of ChainableStack wrappers [here](http://discourse.mc-stan.org/t/wrapping-calls-to-chainablestack/3779) and putting a link in the top of this PR.

**Comment by seantalts** [comment]
And another thread on supporting older Apple Clang releases [here](http://discourse.mc-stan.org/t/potentially-dropping-support-for-older-versions-of-apples-version-of-clang/3780). My proposed path from here might be to get started on the ChainableStack refactoring in another PR (I could take a look at this if you want). I think there's a light version that just replaces a lot of these calls to allocate arrays with the appropriate existing function for that.

**Comment by bob-carpenter** [comment]
I'm always in favor of factoring pull requests if it makes them easier to review.

**Comment by wds15** [comment]
I am the code provider and not the reviewer (so I may judge different as I starred at the code a bit)... but how would the proposed refactor help for reviewing this pull?

I mean, for the matter of this pull we are trading one prefix for another wherever the stack is accessed.

@bob-carpenter: Which function are you referring to which should be used more often to grab memory from the stack? Maybe this light refactoring can just be included in this one if we all think it helps us.

**Comment by bob-carpenter** [comment]
There's a templated alloc_array method that builds in the sizing and casts.  We can just fix all those at once in a separate pull request, though.

**Comment by seantalts** [comment]
Okay, so we can put off refactoring to other issues. But we still have the Apple clang issue being discussed on discourse. In the interest of speed, I might suggest we just ifdef it for now and let the decision to drop compiler support there be orthogonal, as we'd have to rewrite a lot of doc first.

**Comment by wds15** [comment]
I bumped clang on travis to 3.9.1 which hopefully removes the need for this atexit thing. I did chose this version by following the thread here:

https://reviews.llvm.org/D21803?id=70406

where a fix (I think) in this regard was committed on Oct 2016 and 3.9.1 was released on Dec that year.

I am in general fine with if-def-ing this thread_local thing out, but let's have a short discussion today at the meeting. From my perspective this looks like

Pro if-def-out:
- Keep pre-Sierra Apple clangs happy
- certainly clang 3.8 on linux would be easier to handle
- probably a safe choice at this point to if-def-out
- for sure not a performance hit
- easy way to defer decision and move on for now

contra if-def-out
- additional burden on developers: Are we going to provide functions which use threads in two variants?
- additional burden on testing?
- this is vanilla C++11 and we live in 2018
- performance hit likely not there (need to verify; can be done)

**Comment by wds15** [comment]
ok, after struggling a bit more with compilers+travis this is what I think is going on:

- the diff I refer to above did land in clang 4 (and not in 3.9 yet)
- travis has not yet whitelisted the clang 4 repository on the ubuntu we use => we cannot get travis to run our tests with clang 4 unless someone else can fix this

Solutions:
- make travis work with clang 4 and require this as the smallest acceptable clang version (needs to be verified)
- indeed introduce a macro which switches off thread_local => we would still need a clever way to do this automatically (e.g. detect old Apples clangs / clang pre 4.0 / g++ pre 4.8.1)

Sorry for the many changes and posts, but working this out is non-trivial given the compiler/system complexity.

**Comment by wds15** [comment]
ok, clang 4.0 also needs this odd atexit thing. I will stop flipping things around until we either decide to make thread_local optional or I have a good test system close to me.

**Comment by wds15** [comment]
@bgoodri : The problematic configuration for the thread_local stuff is clang on linux. Travis chokes on this configuration with an undefined reference to '__cxa_thread_atexit '. A solution to this is the workaround which I found on stackoverflow and is part of this pull right now (see the `make/tests`).

So I don't like this hack at all as it looks brittle to me and we would also turn this on for cmdstan, etc..

It would be great to find a better solution (maybe a minimal clang version or the requirement to link in some additional library on linux... I keep reading about supc++ but things don't make sense to me and I don't have such a system at my hand). So thanks for looking into this.

**Comment by seantalts** [comment]
I am learning this is a pretty easy use-case for cmake, doing something like is described here: http://www.omniprog.info/clang_bootstrap.html to use cmake to check if the library has the function, and add that file conditional on that.

**Comment by wds15** [comment]
@seantalts : I made the changes as discussed in the meeting. Now it is required that `STAN_THREADS` is defined during compilation. The tests are still running so we will see if the code indeed behaves as I expect. Basically I also figured out a way which will minimize the impact on maintenance for threaded/non-threaded code. If you have a loot at the gradient test you see an example.

Unless @bgoodri comes up with a good solution for clang+linux I would suggest to move forward as:

- add a wiki page to stan-math which describes how to enable threading, under which configs we test this (g++ 4.9) and under which compilers we think it works
- take out the hacks for clang+linux and document them on the wiki

Does that make sense to you? I will update the PR above to note our strategy.

BTW, we never discussed naming, so
- is `STAN_THREADS` a good name for the compiler flag
- is `context()` a good function name? To me it made sense in that this is the thread context, but you could also call this `instance()` or `get_stack()`. I don't have a strong opinion here in either, but if we want to change that name we do it better now.

**Comment by seantalts** [comment]
That all sounds good to me. Maybe we can make a ticket to improve automated threading support in the future (e.g. for clang, detecting if the stdlib being used has that function, etc).

**Comment by bgoodri** [comment]
How do I break this? I have `clang-4.0`. The rev tests work whether I comment out the `__cxa_thread_atexit` thing or not.

**Comment by seantalts** [comment]
The Linux Jenkins box has a recent clang version available, does that help? I don't think we want to switch the default to 4.0, if that's what you're asking - that only came out a year ago.

**Comment by bgoodri** [comment]
That is why I am not seeing what the problem is. The __cxa_thread_atexit
thing was merged into libc++ in 2015, way before clang-4.0 was released.

**Comment by seantalts** [comment]
Damn this unthreaded interface - Sorry, I was replying to Sebastian's question. I am not sure how to break it, though I do know it depends on your libc and not on your clang compiler. So maybe try with the other standard library (clang's if you're on gcc's, vice versa)

**Comment by bgoodri** [comment]
It also does not not work on the computer in my office where `clang++` defaults to version 3.9 and libc++ is apparently version 3.7, according to ( https://packages.ubuntu.com/xenial/libs/libc++1 )

**Comment by wds15** [comment]
@bgoodri I think one would need libc++ 4.0 and clang 4.0 to have a safe bet that it works. Another attempt with clang would be to use it on linux with libstdc++ (the gnu std c++). Could you try that? ...but it looks like we are going for the moment without linux+clang support wrt to the major distributions which I think is fine given that gcc is anyway more common on linux, I think.

**Comment by wds15** [comment]
@seantalts the Linux Jenkins newer clang may work, but given the trouble we are having with this combination we need to check it carefully. Does the Linux Jenkins also have a g++ >= 4.9? That should definitely work with MPI and with threads.

**Comment by wds15** [comment]
Ok, all done. Doc [is here](https://github.com/stan-dev/math/wiki/Threading-Support). I am updating the PR's description as well and will file two issues

- threading in linux & clang
- clean up AD stack access

**Comment by wds15** [comment]
Cool that this is in... map_rect_concurrent is coming soon!

**Comment by bgoodri** [comment]
The tests pass for me on the current develop under Debian stable with clang++-3.8 and the corresponding libc++. It is unlikely that anyone would seriously try to use Stan with something older than that. Although clang++-3.5 technically supports all of C++14, it may not be compatible with libstdc++ >= 4.9.x.

**Comment by wds15** [comment]
Maybe this is a travis problem as they run the super old (2014 released) trusty LTS. If Debian stable is already ok, then I doubt that there is any issue on more recent systems. I would guess that ubuntu xenial (2016) is already OK. However, travis does not seem to have images for that. @bgoodri did u test on xenial?

**Comment by bgoodri** [comment]
Yeah, the server in my office is Xenial, and I haven't been able to trigger
a build failure. Moreover, Bionic Beaver LTS is being released probably
about the same time as Stan 2.18 so Xenial and pre-Xenial will fade away
even more.

**Comment by SteveBronder** [comment]
I didn't see this early enough, but if we wanted to get rid of having to type the  `::context()` in

```cpp
ChainableStack::context().var_stack_.push_back(this);
```

and go back to something like

```cpp
ChainableStack.var_stack_.push_back(this);
```

we could do something like I have below in the OpenCL branch

https://github.com/rok-cesnovar/math/blob/04e09b08ff0073bbaa15ef3305f72d3fcbb721d3/stan/math/gpu/opencl_context.hpp#L53

essentially we do the same Meyer's singleton style that you have here, but then we make a friend class that operates as an API So you end up with functions like 

```cpp
  inline cl::Context& context() {
    return opencl_context_base::getInstance().context_;
  }
```

so eod the user just calls `opencl_context.context()` instead of `opencl_context::instance().context()`

**Pull request merged**